File: color-patterning-with-GA/src/genetic.cpp
//Created by KUAN HAO CHEN
//b04b01047
//Genetic algorithm


#include <algorithm>
//#include <cstring>
#include <string>
#include <fstream>
#include <iostream>
#include <queue>
#include <string>
#include <sstream>
#include <vector>
#include <cstdlib>
#include <cmath>
#include "genetic.h"
#include "shape.h"
using namespace std;



#define CROSSOVER_RATE            0.7
#define MUTATION_RATE             0.001


#define GENE_LENGTH               1
#define MAX_ALLOWABLE_GENERATIONS   400

//returns a float between 0 & 1
#define RANDOM_NUM      ((float)rand()/(RAND_MAX+1.0))
//Constructor
chromo_typ::chromo_typ(int num_group,Coloring_Bounding_Box Box){
	bits = GetRandomBits(num_group);
	fitness = Calculate_fitness(Box);
}


chromo_typ::chromo_typ(int num_group,string bts,Coloring_Bounding_Box Box){
	bits = bts;
	fitness = Calculate_fitness(Box);
}

chromo_typ::chromo_typ(){
	bits = "";
	fitness = 0.0;
}
//		member function

double chromo_typ::Get_fitness(){
	return fitness;
}



string chromo_typ::Get_Bits()
{
	return bits;
}

//
//  This function returns a string of random 1s and 0s of the desired length.
//
//-----------------------------------------------------------------------------------------
string  chromo_typ::GetRandomBits(int length)
{
    string bits;
 
    for (int i=0; i<length; i++)
    {
        if (RANDOM_NUM > 0.5f)
 
            bits += "1";
 
        else
 
            bits += "0";
    }
 
    return bits;
}
 
//---------------------------------BinToDec-----------------------------------------
//
//  converts a binary string into a decimal integer
//
//-----------------------------------------------------------------------------------
int chromo_typ::BinToDec(string bits)
{
    int val          = 0;
    int value_to_add = 1;
 
    for (int i = bits.length(); i > 0; i--)
    {
         
 
        if (bits.at(i-1) == '1')
 
            val += value_to_add;
 
        value_to_add *= 2;
     
    }//next bit
 
    return val;
}
 
 
//---------------------------------Get_fitness------------------------------------------
//According to the area of CA and CB, calculate the fitness
//         It is still unfinished
// 
// 
//------------------------------------------------------------------------------------
double chromo_typ::Calculate_fitness(Coloring_Bounding_Box Box)
{
     Box.color_all_groups(output());

     double score = Box.getScore();

     /*
     for(int i = 0;i<bits.length();i++){
        
        score += stoi(bits.substr(i,1));
     }
    */
    return score;
}




void chromo_typ::AssignFitness(Coloring_Bounding_Box Box)
{
     fitness = Calculate_fitness(Box);
    
}    
void chromo_typ::update(Coloring_Bounding_Box Box){
 	AssignFitness(Box);
 }
//---------------------------------PrintChromo---------------------------------------
//
// decodes and prints a chromo to screen
//-----------------------------------------------------------------------------------
void chromo_typ::PrintChromo()
{   
    //holds decimal values of gene sequence   
         
    int num_elements = bits.length();
     
    for (int i=0; i<num_elements; i++)
    {
        cout<<bits[i];
    }
    cout<<endl;
    
}
vector<int> chromo_typ::output()
{
   vector<int> gene;
   for(int i = 0;i<bits.length();i++){
        
        gene.push_back(stoi(bits.substr(i,1)));
     }
   return gene;
}
//--------------------------------------PrintGeneSymbol-----------------------------
//  
//  given an integer this function outputs its symbol to the screen 
//----------------------------------------------------------------------------------

 

genome::genome(vector<chromo_typ> chromosome,Coloring_Bounding_Box Box,int number_group, int pop_size){
    num_group = number_group;
    POP_SIZE = pop_size;
	chromosomes  = chromosome;
    box = Box;
}
double genome::Calculate_Totalfitness(){
	double TOTAL_FITNESS = 0;
	for(vector<chromo_typ>::iterator chromo = chromosomes.begin(); chromo != chromosomes.end(); ++chromo){
		TOTAL_FITNESS += chromo->Get_fitness();
	}
	return TOTAL_FITNESS;
}


//------------------------------------Mutate---------------------------------------
//
//  Mutates a chromosome's bits dependent on the MUTATION_RATE
//-------------------------------------------------------------------------------------
void genome::Mutate(string &bits)
{
    
    
    for (int i=0; i<bits.length(); i++)
    {
        if (RANDOM_NUM < MUTATION_RATE)
        {
            if (bits.at(i) == '1')
 
                bits.at(i) = '0';
 
            else
 
                bits.at(i) = '1';
        }
    }
    
    return;
}
 
//---------------------------------- Crossover ---------------------------------------
//
//  Dependent on the CROSSOVER_RATE this function selects a random point along the 
//  lenghth of the chromosomes and swaps all the  bits after that point.
//------------------------------------------------------------------------------------
void genome::Crossover(string &offspring1, string &offspring2)
{
  
  //dependent on the crossover rate
  if (RANDOM_NUM < CROSSOVER_RATE)
  {
    
    //create a random crossover point
    int crossover = (int) (RANDOM_NUM * (num_group-1));
    
    string t1 = offspring1.substr(0, crossover) + offspring2.substr(crossover, num_group);
    string t2 = offspring2.substr(0, crossover) + offspring1.substr(crossover, num_group);
    
    offspring1 = t1; offspring2 = t2;                 
  }
  
}
 
 
//--------------------------------Roulette-------------------------------------------
//
//  selects a chromosome from the population via roulette wheel selection
//------------------------------------------------------------------------------------
string genome::Roulette(int total_fitness, vector<chromo_typ> Population)
{
    //generate a random number between 0 & total fitness count
    float Slice = (float)(RANDOM_NUM * total_fitness);
     
    //go through the chromosones adding up the fitness so far
    float FitnessSoFar = 0.0f;
     
    for (int i=0; i<POP_SIZE; i++)
    {
        FitnessSoFar += Population[i].Get_fitness();
         
        //if the fitness so far > random number return the chromo at this point
        if (FitnessSoFar >= Slice)
            
            return Population[i].Get_Bits();
    }
 
    return "";
}






void genome::reproduce(){
	total_fitness = Calculate_Totalfitness();
	vector<chromo_typ> temp;
    int cPop = 0;
    
    //loop until we have created POP_SIZE new chromosomes
    while (cPop < POP_SIZE)
    {
        
        // we are going to create the new population by grabbing members of the old population
        // two at a time via roulette wheel selection.
        string offspring1 = Roulette(total_fitness, chromosomes);
        string offspring2 = Roulette(total_fitness, chromosomes);
        
        //add crossover dependent on the crossover rate
        Crossover(offspring1, offspring2);
        
        //now mutate dependent on the mutation rate
        Mutate(offspring1);
        Mutate(offspring2);
        
        //add these offspring to the new population. (assigning zero as their
        //fitness scores)
        
        temp.push_back(chromo_typ(num_group,offspring1,box));
        cPop++;
        temp.push_back(chromo_typ(num_group,offspring2,box));
        cPop++;
    }//end loop
    
    //copy temp population into main population array
    chromosomes.clear();
    for (int i=0; i<POP_SIZE; i++)
    {
        chromosomes.push_back(temp[i]);
    }

}
void genome::evolve(){
	
    for(vector<chromo_typ>::iterator chromo = chromosomes.begin(); chromo != chromosomes.end(); chromo++){
		chromo->update(box);
	}
    
	reproduce();
    
}
chromo_typ genome::GetBestGene(){
    double BestFitness = 0.0;
    int Best;
    for(int i = 0;i<chromosomes.size();i++){
        if (chromosomes[i].Get_fitness() > BestFitness){
            BestFitness = chromosomes[i].Get_fitness();
            Best = i;
        }
    }
    return chromosomes[Best];

}

File: genetic.py
# 遗传算法
import random
import copy

# 定义全局变量
cross_rate = 0.74        # 交叉率
mutation_rate = 0.002   # 变异率



# 个体类
class CHROMOSOME:
    def __init__(self, case, length=0, bits=""):
        self.bits = ''
        self.fitness = 0
        
        if len(bits) == 0:
            self.gen_random_bits(length)
        else:
            self.bits = bits

        self.cal_fitness(case)

    def get_bits(self): 
        return self.bits
    def get_fitness(self):
        return self.fitness
    
    # 随机生成个体
    def gen_random_bits(self, length):
        for i in range(0, length):
            if random.random() > 0.5:
                self.bits += "1"
            else:
                self.bits += "0"

    # 计算当前个体的适应度
    def cal_fitness(self, case):
        # 根据当前的个体上色
        case.color_all_groups(self.bits)
        # 获取当前的上色结果的评分作为适应度
        self.fitness = case.cal_score()

# 种群类
class GENOME:

    def __init__(self, chromosomes, case, chromosome_length, chromosome_num):
        self.chromosomes = chromosomes
        self.case = case
        self.chromosome_length = chromosome_length  
        self.chromosome_num = chromosome_num         
        self.total_fitness = 0

    # 计算整体的适应度
    def cal_total_fitness(self):
        self.total_fitness = 0
        for chromosome in self.chromosomes:
            self.total_fitness += chromosome.get_fitness()

    # 变异
    def mutate(self, chromosome_bits):
        global mutation_rate
        chromosome_bits_temp = []
        for bit  in chromosome_bits:
            if random.random() < mutation_rate:
                if bit == '0':
                    chromosome_bits_temp.append('1')
                else:
                    chromosome_bits_temp.append('0')
            else:
                chromosome_bits_temp.append(bit)
        
        return ''.join(chromosome_bits_temp)

    # 交叉
    def crossover(self, chromosome_bits_1, chromosome_bits_2):
        global cross_rate
        if random.random() < cross_rate:
            # 随机产生交叉点
            crossover_point = random.randint(0, self.chromosome_length-1)
            t1 = chromosome_bits_1[0:crossover_point] + chromosome_bits_2[crossover_point:]
            t2 = chromosome_bits_2[0:crossover_point] + chromosome_bits_1[crossover_point:]
            
            return [t1, t2]

        return [chromosome_bits_1, chromosome_bits_2]

    # 筛选个体（轮盘法，roulette wheel selection）
    def roulette_wheel_selecte(self):
        slice = random.random() * self.total_fitness
        fitness_so_far = 0
        for chromosome in self.chromosomes:
            fitness_so_far += chromosome.get_fitness()
            if(fitness_so_far >= slice):
                return chromosome.get_bits()
    
    # 获取当前种群最佳的个体
    def get_best_chromosome(self):
        best_fitness = 0
        best_index = 0
        for index, chromosome in enumerate(self.chromosomes):
            if chromosome.get_fitness() > best_fitness:
                best_fitness = chromosome.get_fitness()
                best_index = index

        return self.chromosomes[best_index]

    # 生成新的种群
    def gen_new_chromosome(self):
        # 计算当前种群总的适应度
        self.cal_total_fitness()
        # 新的种群
        chromosomes_temp = []
        chromosomes_count = 0
        
        while chromosomes_count < self.chromosome_num:

            bits_1 = self.roulette_wheel_selecte()
            bits_2 = self.roulette_wheel_selecte()


            [bits_1, bits_2] = self.crossover(bits_1, bits_2)


            bits_1 = self.mutate(bits_1)
            bits_2 = self.mutate(bits_2)


            chromosomes_temp.append(CHROMOSOME(self.case, bits=bits_1))
            chromosomes_temp.append(CHROMOSOME(self.case, bits=bits_2))
            
            chromosomes_count += 2
        

        self.chromosomes = copy.deepcopy(chromosomes_temp)
